src/Azure/AzureStorage.php

```diff
@@ -0,0 +1,315 @@
+    {
+        $canonicalizedResource = $this->getCanonicalizedResourcePath($filename, $destinationPath);
+
+        $url = 'https://management.azure.com/subscriptions/'
+            . $this->config['subscriptionId'] . '/resourceGroups/' . $this->config['resourceGroup']
+            . '/providers/Microsoft.Storage/storageAccounts/' . $this->config['storageAccount']
+            . '/listServiceSas?api-version=' . $this->config['apiVersion'];
+
+        try {
+            $response = $this->makeSasRequest($url, $canonicalizedResource);
+
+            if ($response->getStatusCode() >= 400) {
+                throw new \Exception();
+            }
+        } catch (\Exception $e) {
+            $this->getAccessToken(true);
+            $response = $this->makeSasRequest($url, $canonicalizedResource);
+        }
+
+        $body = $response->getContent(false);
+        $data = json_decode($body, true);
+
+        if (!array_key_exists('serviceSasToken', $data)) {
+            throw new \Exception('Failed generating SAS signature');
+        }
+
+        return $data['serviceSasToken'];
+    }
+
+    /**
+     * Make the POST request to Azure for getting a SAS token
+     *
+     *
+     * @param string $url
+     * @param string $canonicalizedResource
+     *
+     * @return ResponseInterface
+     *
+     * @throws \Psr\Cache\InvalidArgumentException
+     * Can be thrown deep inside the Cache interface, see Symfony\Component\Cache\Traits\ConstraintsTrait:63
+     *
+     * @throws \Symfony\Contracts\HttpClient\Exception\TransportExceptionInterface
+     */
+    private function makeSasRequest(string $url, string $canonicalizedResource): ResponseInterface
+    {
+        return $this->client->request(
+            'POST',
+            $url,
+            [
+                RequestOptions::HEADERS => [
+                    'Authorization' => 'Bearer ' . $this->accessToken,
+                    'Content-Type' => 'application/json',
+                ],
+                RequestOptions::JSON => [
+                    'canonicalizedResource' => $canonicalizedResource,
+                    'signedResource' => $this->config['signedResource'],
+                    'signedPermission' => $this->config['permissions'],
+                    'signedProtocol' => 'https',
+                    'signedExpiry' => $this->config['expiry'],
+                    'signedStart' => $this->config['start'],
+                ],
+            ]
+        );
+    }
+
+    /**
+     * Get the URL for accessing a file
+     *
+     * @param string $filename
+     * The file name
+     *
+     * @param string $signature
+     * The signature for accessing this file
+     *
+     * @param string|null $destinationPath
+     * Optional: the path where the file is be stored
+     *
+     *
+     * @return string
+     * The URL to access the file using the SAS signature
+     */
+    private function getFileUrl(string $filename, string $signature, ?string $destinationPath = null): string
```

Comment:
Je zou kunnen overwegen $signature optioneel te maken, en de signature in deze functie op te halen met behulp van de $filename en $destinationPath. Dan kun je het zo maken dat je aan `getFileUrl` alleen maar een $filename hoeft mee te geven. Dan kun je uiteindelijk in je public function dit:

```


        $signature = $this->generateSASSignature($filename, $destinationPath);

        // Create the signed blob URL
        return $this->getFileUrl($filename, $signature, $destinationPath);
```

simplificeren naar dit:

```
return $this->getFileUrl($filename, $destinationPath);
```

Reply:
Slimmerik! Na het lezen was ik het er mee eens en heb ik hem aangepast!